### ISSUE TITLE
Documentation Issue/ or software Issue at #_connecting-to-multiple-containers-at-once

### ISSUE THREAD

**Issue by Waelder** [microsoft/vscode-remote-release#4885]
Hi,

 want have a multi container app ( db, python, react).
I want to run it locally in docker, and connect to the python as well as to the react container. 
The last obstacle is the automatically generated "compose project name" which is different.

Following the documentation i created a .env file to create a name for docker compose.
otherwise the 2nd. open folder in container would fire up a complete new compose stack.

Unfortunately when i follow this documentation, i had to create copy of .env file  in the folder wich i opened with the 2nd windows in the container. Probably because then the subfolder is then considered as project root.

Kind regards

**Comment by chrmarti** [comment]
Adding support for `--env-file` would solve your request from what I see. (Updating the issue's title.)

**Comment by Waelder** [comment]
yes, sounds good, however the documentation suggest that it "just somehow magically works" 
but with the --env-file at least we can achieve the goal.but it would require a documenation update as well, for other newbies like me.

**Comment by PONzu-0529** [comment]
I have a similar problem.

I'm using Docker Desktop, when using Docker Compose V2, I had to change the location of `.env`.
It is possible to load `.env` in the container with `docker-compose.yml`, but it cannot load when `docker-compose` is ran.

So, I would like to be able to load `.env` at build in the `devcontainer.json`.

**Comment by fyyyyy** [comment]
I want to load an .env file at build time without docker-compose, this seems not possible atm ?

**Comment by chrmarti** [comment]
@fyyyyy Correct, we don't have the env-file option yet and, unlike docker-compose, the docker CLI does not read the `.env` file from the current folder by default.

**Comment by fyyyyy** [comment]
There are 2 options to pass env variables in the docker cli. One is via --env, the other via --env-file
Passing those via --env seems to work with devContainers at BUILD time. However passing an --env-file in runArgs makes them available AFTER the build. This difference of outcomes confuses me.

`docker run -e MYVAR1 --env MYVAR2=foo --env-file ./env.list ubuntu bash`

**Comment by fyyyyy** [comment]
I have found a workaround to set ENV vars from a file during the docker build. This is i.e. necessary if you want to use a private NPM registry during build, but you don't want to write the secrets in devcontainer.json as this file is comitted to the repo.

devcontainer.env ( add this to .gitignore )
```
export NPM_USER=***
export NPM_PASS=***
export NPM_EMAIL=***
```

In the Dockerfile:
```
RUN npm i -g --production --registry https://registry.npmjs.org/ npm-cli-adduser

# Set env variables from devcontainer.env. Needs to be in same RUN step as env vars will be lost in next RUN
COPY devcontainer.env devcontainer.env
RUN $(cat devcontainer.env) && npm-cli-adduser -r *private-registry-url* \
&& npm whoami

# install private registry package
RUN npm i -g  -r *private-registry-url* *private-registry-package*
```

**Comment by bgloss** [comment]
https://github.com/orgs/devcontainers/discussions/36

**Comment by bgloss** [comment]
vscode executes `docker compose` from the folder of the first `docker-compose.yml` file in the list.

When you place an empty `docker-compose.yml` into the root folder of your projects and list it as the first `docker-compose.yml` to use, you get what you want. 

Two nice side-effects:
- The .env file is loaded from the root of the projects (which you cannot control otherwise)
- The namespace that docker uses is that of the root folder and thus same for all sub-projects

my `devontainer.json` looks like this:
```json
	"dockerComposeFile": [
		"../../docker-compose.yml",
		"../../environments/dev_local/postgres/docker-compose.yml",
		"./docker-compose.yml"
	],
```

The docker-compose.yml in the root folder is this:
```yaml
version: '3.8'
# intentionally empty
```

What makes it a bit quirky is that the paths of the `docker-compose` files are relative to the project and the paths inside the `docker-compose.yml` files are relative to the root of the projects.

It is just a work-around. I think a proper solution is needed in `devcontainer.json` to mange it.

**Comment by bgloss** [comment]
To add to the feature request ... I think just adding a method to provide the `--env-file` option into the `docker compose` scenario is not enough. 

As it is already established for the ["Image and Dockerfile Scenario"](https://containers.dev/implementors/json_reference/#image-specific)  a similar approach to control and externalize key settings in the ["Docker Compose Scenario"](https://containers.dev/implementors/json_reference/#compose-specific) would be needed. 

Just for comparison: This is how a `devcontainer.json` looks like when using the Dockerfile scenario:

```json
	"build": {
		"dockerfile": "Dockerfile",
		"context": "..",
		"args": { 
			"VARIANT": "3.10-bullseye",
			"NODE_VERSION": "lts/*"
		},
	},
        "runArgs": [ "--env-file=../.env" ]
```
similar methods to specifiy `context`, `args` and `runArgs` would be needed.

**Comment by metaskills** [comment]
So my needs align more with (https://github.com/microsoft/vscode-remote-release/issues/5324) which was closed as a duplicate and now this issues title "using --env-file" may not. Let me explain.

I would like our base .devcontainer folder to work for Codespaces and our CI/CD via GitHub Actions (using devcontainers/ci). As such, the devcontainer.json would leverage a `"dockerComposeFile": ["docker-compose.yml"]` and a key part of the compose service would be an Oracle DB. This DB does not work on arm64 (Apple Silicon) and there is a need to run our devcontainers locally. I would like a way for users to configure a non git committed file to remove the Oracle DB service. My hope was something like this:

1. Commit a .devcontainer/docker-compose-local.yml file. (contents below)
2. Git ignore the .devcontiner/.env file.
3. Have Mac users create the .devcontiner/.env with a COMPOSE_FILE env variable (seen below)

```yaml
# .devcontainer/docker-compose-local.yml
services:
  oracle:
    image: hello-world
    restart: "no"
    volumes: []
```

```bash
# .devcontiner/.env
COMPOSE_FILE=docker-compose.yml:docker-compose-local.yml
```

Would the solutions proposed here help me with that or should we open this back up?

- https://github.com/microsoft/vscode-remote-release/issues/5324

**Comment by gislerro** [comment]
I'd suggest to extend the [devcontainer compose spec](https://containers.dev/implementors/json_reference/#compose-specific) to include a `dockerComposeArgs` option, which are then passed to the `docker compose` command:

```json
{
    "dockerComposeFile": ["a.compose.yaml", "b.compose.yaml"],
    "dockerComposeArgs": ["--env-file custom.env"]
}
```

should result in `docker compose -f a.compose.yaml -f b.compose.yaml --env-file custom.env`

So the docker compose files will be interpolated based on `custom.env`.

I guess some [compose args](https://docs.docker.com/reference/cli/docker/compose/#options) may interfere with how vscode sets up the container or just don't make sense (`--dry-run` for example). So allowing a subset of the args would be ideal

**Comment by TeaDrinkingProgrammer** [comment]
Any updates on this?

**Comment by luketych** [comment]
ANy updates?

**Comment by SKairinos** [comment]
Any updates?

**Comment by thomasbival** [comment]
Any updates?

**Comment by mrmachine** [comment]
Please implement `--env-file` support for docker compose. I've just spent hours trying to make my project run in dev container that needs a different .env file for docker compose than when run outside dev container

**Comment by RaphvK** [comment]
One possible workaround is to add another compose file, e.g., `docker-compose.dev.yml` to include the actual compose file `docker-compose.yml` setting the env file and use this file in `devcontainer.json`.

```yaml
# docker-compose.dev.yml
include:
  - path: docker-compose.yml
    env_file: .env
```

**Comment by galshi** [comment]
Hi, Any updates on this feature?

**Comment by Xentox-Phil** [comment]
yeah any updates on this! this would be amazing!

**Comment by andrian2929** [comment]
Any update on this?

**Comment by hookenz** [comment]
It's a glaring hole and makes devcontainers quite inflexible.  I want to pass environment vars to ALL services that.    dockerComposeArgs would be great.   I could then add things like "--profile sqlite" as well.

I have a number of different dev container configurations for one project.  Because of the lack of env file override support I need to create 6 different docker-compose files to support the various configurations.   It's painful and error prone.  
But I also want to pass profile.
